File: assets/reports-model/R/Donation Frequency.R
# ============================================
# Donation Frequency Interactive Pie Chart
# ============================================

set.seed(123)  # For reproducible sample data
# Sample donation frequency distribution
sample_frequency <- c(
  rep("1st Time", 428),
  rep("Repeat", 526)
)

# Shuffle for randomness
sample_frequency <- sample(sample_frequency)

# Create a data frame (this simulates your donor_form data)
donor_data_raw <- data.frame(
  donor_id = 1:length(sample_frequency),
  donation_frequency = sample_frequency
)

# ============================================
# TO USE YOUR ACTUAL DATA:
# Read your data from CSV or other source
# ============================================
# Option 1: If you have a CSV file
# donor_data_raw <- read.csv("donor_form.csv")

# Option 2: If you have an Excel file
# install.packages("readxl")
# library(readxl)
# donor_data_raw <- read_excel("donor_form.xlsx")

# Option 3: If you manually input data
# donor_data_raw <- data.frame(
#   donor_id = c(1, 2, 3, ...),
#   donation_frequency = c("1st Time", "Repeat", ...)
# )

# ============================================
# Data Processing - Count by Donation Frequency
# ============================================
donor_data <- donor_data_raw %>%
  group_by(donation_frequency) %>%
  summarise(count = n()) %>%
  mutate(
    percentage = round((count / sum(count)) * 100, 1),
    donation_frequency = factor(donation_frequency, 
                                levels = c("1st Time", "Repeat"))
  ) %>%
  arrange(donation_frequency)

# Total donors
total_donors <- sum(donor_data$count)

# ============================================
# Create Interactive Pie Chart
# ============================================
# Define color palette - light blue for 1st time, deep blue for repeat
colors <- c('#60a5fa', '#2563eb')

# Create the plot
fig <- plot_ly(
  data = donor_data,
  labels = ~donation_frequency,
  values = ~count,
  type = 'pie',
  hole = 0.4,
  marker = list(
    colors = colors,
    line = list(color = '#ffffff', width = 2)
  ),
  hovertext = ~paste0(
    'Frequency: ', donation_frequency, '<br>',
    'Count: ', count, '<br>',
    'Percentage: ', percentage, '%'
  ),
  hoverinfo = 'text',
  textposition = 'inside',
  textinfo = 'none',
  hole = 0.3  # Creates a donut chart (set to 0 for full pie)
) %>%
  layout(
    title = list(
      text = paste0(
        '<b>Donation Frequency Distribution</b><br>',
        '<sub>Total Donors: ', total_donors, '</sub>'
      ),
      font = list(size = 20)
    ),
    showlegend = TRUE,
    legend = list(
      orientation = 'v',
      x = 1,
      y = 0.5,
      font = list(size = 12)
    ),
    paper_bgcolor = '#ffffff',
    plot_bgcolor = '#ffffff',
    hoverlabel = list(
      bgcolor = 'white',
      font = list(size = 12, color = 'black'),
      bordercolor = '#e5e7eb'
    ),
    margin = list(t = 80, b = 40, l = 40, r = 150)
  ) %>%
  config(
    displayModeBar = TRUE,
    modeBarButtonsToRemove = c('lasso2d', 'select2d'),
    displaylogo = FALSE
  )

# Display the chart
fig

# ============================================
# Print Summary Statistics
# ============================================
cat("\n=== DONATION FREQUENCY SUMMARY ===\n\n")
cat(sprintf("Total Donors: %d\n\n", total_donors))

for(i in 1:nrow(donor_data)) {
  cat(sprintf(
    "%s Donors: %d (%.1f%%)\n",
    donor_data$donation_frequency[i],
    donor_data$count[i],
    donor_data$percentage[i]
  ))
}

# ============================================
# Additional Insights
# ============================================
first_time_count <- donor_data %>% 
  filter(donation_frequency == "1st Time") %>% 
  pull(count)

repeat_count <- donor_data %>% 
  filter(donation_frequency == "Repeat") %>% 
  pull(count)

retention_rate <- round((repeat_count / total_donors) * 100, 1)

cat("\n=== DONOR RETENTION INSIGHTS ===\n\n")
cat(sprintf("First-Time Donors: %d\n", first_time_count))
cat(sprintf("Repeat Donors: %d\n", repeat_count))
cat(sprintf("Retention Rate: %.1f%%\n", retention_rate))

if(repeat_count > first_time_count) {
  cat("\n✓ More repeat donors than first-time donors - Good retention!\n")
} else {
  cat("\n⚠ More first-time donors - Focus on retention strategies.\n")
}

File: assets/reports-model/Donation Frequency.py
"""
Donation Frequency Distribution
===============================
Creates an interactive pie/donut chart showing donation frequency
distribution for ACTIVE donors only.

Definition for Reports dashboard:
- Use the eligibility history table
- Consider ONLY donors whose latest eligibility.status is APPROVED
- For those active donors:
    - "1st Time"  -> only one APPROVED record in history
    - "Repeat"    -> two or more APPROVED records in history
"""

import json
from collections import defaultdict
from typing import Dict, List

import plotly.graph_objects as go

from database import DatabaseConnection


def _parse_dt(value):
    from datetime import datetime

    if not value:
        return None
    try:
        text = str(value)
        if "Z" in text:
            return datetime.fromisoformat(text.replace("Z", "+00:00"))
        return datetime.fromisoformat(text)
    except Exception:
        return None


def fetch_eligibility_history() -> List[Dict]:
    """
    Fetch full eligibility history ordered by donor_id and created_at.
    """
    db = DatabaseConnection()
    records: List[Dict] = []

    if db.connect():
        try:
            endpoint = (
                "eligibility?"
                "select=eligibility_id,donor_id,status,created_at"
                "&order=donor_id.asc,created_at.asc"
            )
            records = db.supabase_request(endpoint)
        finally:
            db.disconnect()

    return records


def aggregate_donation_frequency(records: List[Dict]) -> Dict:
    """
    Aggregate ACTIVE donors by donation frequency using eligibility history.

    Returns:
        {
          "data": [
             { donation_frequency: "1st Time", count, percentage },
             { donation_frequency: "Repeat",   count, percentage },
          ],
          "total_donors": N
        }
    """
    # Track all statuses and approvals per donor
    history_by_donor: Dict[int, List[Dict]] = defaultdict(list)

    for rec in records:
        donor_id = rec.get("donor_id")
        if donor_id is None:
            continue
        history_by_donor[donor_id].append(rec)

    first_time = 0
    repeat = 0

    for donor_id, events in history_by_donor.items():
        # Determine latest status
        latest = max(events, key=lambda r: _parse_dt(r.get("created_at")) or 0)
        status_text = (latest.get("status") or "").lower()

        # Only consider donors whose latest status is currently approved
        if not ("approved" in status_text or status_text == "eligible"):
            continue

        # Count approvals across full history
        approvals = sum(
            1
            for r in events
            if "approved" in (r.get("status") or "").lower()
        )

        if approvals <= 1:
            first_time += 1
        else:
            repeat += 1

    total = first_time + repeat
    if total == 0:
        return {"data": [], "total_donors": 0}

    data = [
        {
            "donation_frequency": "1st Time",
            "count": first_time,
            "percentage": round((first_time / total) * 100, 1),
        },
        {
            "donation_frequency": "Repeat",
            "count": repeat,
            "percentage": round((repeat / total) * 100, 1),
        },
    ]

    return {"data": data, "total_donors": total}


def create_pie_chart(donor_data: List[Dict], total_donors: int) -> go.Figure:
    """Create an interactive pie/donut chart using Plotly."""
    # Color palette - light blue for 1st time, deep blue for repeat
    colors = ["#60a5fa", "#2563eb"]
    
    labels = [d["donation_frequency"] for d in donor_data]
    values = [d["count"] for d in donor_data]
    
    # Create hover text
    hover_text = [
        f"Frequency: {d['donation_frequency']}<br>Count: {d['count']}<br>Percentage: {d['percentage']}%"
        for d in donor_data
    ]
    
    fig = go.Figure(data=[
        go.Pie(
            labels=labels,
            values=values,
            hole=0.3,  # Donut chart
            marker=dict(
                colors=colors,
                line=dict(color="#ffffff", width=2)
            ),
            hovertext=hover_text,
            hoverinfo="text",
            textposition="inside",
            textinfo="none"
        )
    ])
    
    fig.update_layout(
        title=dict(
            text=f"<b>Donation Frequency Distribution</b><br><sub>Total Donors: {total_donors}</sub>",
            font=dict(size=20)
        ),
        showlegend=True,
        legend=dict(
            orientation="v",
            x=1,
            y=0.5,
            font=dict(size=12)
        ),
        paper_bgcolor="#ffffff",
        plot_bgcolor="#ffffff",
        hoverlabel=dict(
            bgcolor="white",
            font_size=12,
            font_color="black",
            bordercolor="#e5e7eb"
        ),
        margin=dict(t=80, b=40, l=40, r=150)
    )
    
    return fig


def get_donation_frequency_data() -> Dict:
    """
    Main function to get donation frequency data.
    Returns a dictionary with the data for API consumption.
    """
    records = fetch_eligibility_history()

    if not records:
        return {
            "success": False,
            "message": "No eligibility records found in database",
            "data": [],
            "total_donors": 0,
            "retention_rate": 0,
        }

    result = aggregate_donation_frequency(records)
    donor_data = result["data"]
    total_donors = result["total_donors"]

    # Calculate retention rate
    repeat_count = next(
        (d["count"] for d in donor_data if d["donation_frequency"] == "Repeat"), 0
    )
    retention_rate = round((repeat_count / total_donors) * 100, 1) if total_donors > 0 else 0

    return {
        "success": True,
        "data": donor_data,
        "total_donors": total_donors,
        "retention_rate": retention_rate,
    }


def generate_chart_html(output_path: str = None) -> str:
    """Generate and optionally save the chart as HTML."""
    records = fetch_eligibility_history()

    if not records:
        print("WARNING: No eligibility records found in database", file=sys.stderr)
        return ""

    result = aggregate_donation_frequency(records)
    donor_data = result["data"]
    total_donors = result["total_donors"]

    fig = create_pie_chart(donor_data, total_donors)

    if output_path:
        fig.write_html(output_path)
        print(f"Chart saved to: {output_path}")

    return fig.to_html(include_plotlyjs="cdn")


def print_summary():
    """Print a formatted summary to console (matching R output style)."""
    result = get_donation_frequency_data()
    
    print("\n=== DONATION FREQUENCY SUMMARY ===\n")
    print(f"Total Donors: {result['total_donors']}\n")
    
    for item in result["data"]:
        print(f"{item['donation_frequency']} Donors: {item['count']} ({item['percentage']}%)")
    
    # Additional insights
    first_time_count = next((d["count"] for d in result["data"] if d["donation_frequency"] == "1st Time"), 0)
    repeat_count = next((d["count"] for d in result["data"] if d["donation_frequency"] == "Repeat"), 0)
    
    print("\n=== DONOR RETENTION INSIGHTS ===\n")
    print(f"First-Time Donors: {first_time_count}")
    print(f"Repeat Donors: {repeat_count}")
    print(f"Retention Rate: {result['retention_rate']}%")
    
    if repeat_count > first_time_count:
        print("\n✓ More repeat donors than first-time donors - Good retention!")
    else:
        print("\n⚠ More first-time donors - Focus on retention strategies.")


if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description="Donation Frequency Distribution Report")
    parser.add_argument("--json", action="store_true", help="Output as JSON instead of formatted text")
    parser.add_argument("--html", type=str, help="Save chart as HTML to specified path")
    args = parser.parse_args()
    
    if args.html:
        generate_chart_html(args.html)
    elif args.json:
        result = get_donation_frequency_data()
        print(json.dumps(result, indent=2))
    else:
        print_summary()

